php_phongo.c

```diff
@@ -2841,18 +2841,379 @@ static bool phongo_manager_set_auto_encryption_opts(php_phongo_manager_t* manage
+		mongoc_client_encryption_opts_destroy(opts);
+	}
+} /* }}} */
+
+static mongoc_client_encryption_datakey_opts_t* phongo_clientencryption_datakey_opts_from_zval(zval* options TSRMLS_DC) /* {{{ */
+{
+	mongoc_client_encryption_datakey_opts_t* opts;
+
+	opts = mongoc_client_encryption_datakey_opts_new();
+
+	if (!options || Z_TYPE_P(options) != IS_ARRAY) {
+		return opts;
+	}
+
+	if (php_array_existsc(options, "keyAltNames")) {
+		zval*      zkeyaltnames = php_array_fetchc(options, "keyAltNames");
+		HashTable* ht_data;
+		uint32_t   keyaltnames_count;
+		char**     keyaltnames;
+		uint32_t   i      = 0;
+		bool       failed = false;
+
+		if (!zkeyaltnames || Z_TYPE_P(zkeyaltnames) != IS_ARRAY) {
+			phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Expected keyAltNames to be array, %s given", PHONGO_ZVAL_CLASS_OR_TYPE_NAME_P(zkeyaltnames));
+			goto cleanup;
+		}
+
+		ht_data           = HASH_OF(zkeyaltnames);
+		keyaltnames_count = ht_data ? zend_hash_num_elements(ht_data) : 0;
+		keyaltnames       = bson_malloc(sizeof(char*) * keyaltnames_count);
+
+#if PHP_VERSION_ID >= 70000
+		{
+			zval* keyaltname;
+
+			ZEND_HASH_FOREACH_VAL(ht_data, keyaltname)
+			{
+				if (i == keyaltnames_count) {
+					phongo_throw_exception(PHONGO_ERROR_LOGIC TSRMLS_CC, "Iterating over too many keyAltNames. Please file a bug report");
+					failed = true;
+				}
+
+				keyaltnames[i] = bson_strdup(Z_STRVAL_P(keyaltname));
+				i++;
+			}
+			ZEND_HASH_FOREACH_END();
+		}
+#else
+		{
+			HashPosition pos;
+			zval**       keyaltname;
+			bool         failed = false;
+
+			for (
+				zend_hash_internal_pointer_reset_ex(ht_data, &pos);
+				zend_hash_get_current_data_ex(ht_data, (void**) &keyaltname, &pos) == SUCCESS;
+				zend_hash_move_forward_ex(ht_data, &pos)) {
+
+				if (i == keyaltnames_count) {
+					phongo_throw_exception(PHONGO_ERROR_LOGIC TSRMLS_CC, "Iterating over too many keyAltNames. Please file a bug report");
+					failed = true;
+					break;
+				}
+
+				keyaltnames[i] = bson_strdup(Z_STRVAL_PP(keyaltname));
+				i++;
+			}
+		}
+#endif
+
+		if (!failed) {
+			mongoc_client_encryption_datakey_opts_set_keyaltnames(opts, keyaltnames, keyaltnames_count);
+		}
+
+		for (i = 0; i < keyaltnames_count; i++) {
+			bson_free(keyaltnames[i]);
+		}
+		bson_free(keyaltnames);
+
+		if (failed) {
+			goto cleanup;
+		}
+	}
+
+	if (php_array_existsc(options, "masterKey")) {
+		bson_t masterkey = BSON_INITIALIZER;
+
+		php_phongo_zval_to_bson(php_array_fetchc(options, "masterKey"), PHONGO_BSON_NONE, &masterkey, NULL TSRMLS_CC);
+		if (EG(exception)) {
+			goto cleanup;
+		}
+
+		mongoc_client_encryption_datakey_opts_set_masterkey(opts, &masterkey);
+	}
+
+	return opts;
+
+cleanup:
+	if (opts) {
+		mongoc_client_encryption_datakey_opts_destroy(opts);
+	}
+
+	return NULL;
+} /* }}} */
+
+void phongo_clientencryption_create_datakey(php_phongo_clientencryption_t* clientencryption, zval* return_value, char* kms_provider, zval* options TSRMLS_DC) /* {{{ */
```

Comment:
AFAIK, this function is only called from a method in `phongo_clientencryption_create_datakey`. Perhaps that's the same case for a few other functions (e.g. encrypt, decrypt, and their options helpers). Would it make sense to define these as static functions in ClientEncryption.c?

Perhaps the ClientEncryption class itself doesn't even need to exist (at runtime) unless FLE is enabled. That might be feasible if you wrap most of ClientEncryption.c itself in an `#ifdef MONGOC_ENABLE_CLIENT_SIDE_ENCRYPTION` (after including the necessary config header) and do the same for the `php_phongo_clientencryption_init_ce` call in `PHP_MINIT_FUNCTION(mongodb)` later in this file.

Reply:
> AFAIK, this function is only called from a method in `phongo_clientencryption_create_datakey`. Perhaps that's the same case for a few other functions (e.g. encrypt, decrypt, and their options helpers). Would it make sense to define these as static functions in ClientEncryption.c?

I've put the functions into php_phongo.c to not pollute the class file with unrelated methods. We had previously discussed keeping those files clean with class-related code only. I have no problem moving the functions, but was also planning on re-organising code across the board with the goal to clean up the php_phongo file.

> Perhaps the ClientEncryption class itself doesn't even need to exist (at runtime) unless FLE is enabled.

I'm not sure that's a good idea. The class can't be instantiated by itself, and with the `Manager::clientEncryption` method throwing an exception if FLE is not compiled, the class can't be used. I prefer that over the class only being available in some instances, no matter how useless it may be without libmongocrypt.